src/core/Proc/Async.pm6

```diff
@@ -272,16 +272,16 @@ my class Proc::Async {
     }
 
     method start(Proc::Async:D: :$scheduler = $*SCHEDULER, :$ENV, :$cwd = $*CWD) {
-        X::Proc::Async::AlreadyStarted.new(proc => self).throw if $!started;
-        $!started = True;
-
-        my @blockers;
-        if $!stdin-fd ~~ Promise {
-            @blockers.push($!stdin-fd.then({ $!stdin-fd := .result }));
+        when cas($!started, 0, 1) == 0 {
+            my @blockers;
+            if $!stdin-fd ~~ Promise {
+                @blockers.push($!stdin-fd.then({ $!stdin-fd := .result }));
+            }
+            @blockers
+                ?? start { await @blockers; await self!start-internal(:$scheduler, :$ENV, :$cwd) }
+                !! self!start-internal(:$scheduler, :$ENV, :$cwd)
         }
-        @blockers
-            ?? start { await @blockers; await self!start-internal(:$scheduler, :$ENV, :$cwd) }
-            !! self!start-internal(:$scheduler, :$ENV, :$cwd)
+        X::Proc::Async::AlreadyStarted.new(proc => self).throw if $!started;
```

Comment:
Won't this always throw an exception? Additionally I think you'd want to do an atomic read of `$!started` ( although I'm not sure you need an `atomicint` at all for this )

Reply:
Since the ```when``` block is used, it won't execute the following statements in the same outer block. It looks bit cleaner than an if/else chain IMO.

Reply:
> It looks bit cleaner

Note that it's also about an order of magnitude slower:

```
<Zoffix__> c: HEAD sub foo ($x) { when $x == 1 { }; Nil }; my $x = 1; foo $x for ^10_000_000; say now - ENTER now
<committable6> Zoffix__, ¦HEAD(1ea3ff2): «5.9091969␤»
<Zoffix__> c: HEAD sub foo ($x) { if $x == 1 { } else {} }; my $x = 1; foo $x for ^10_000_000; say now - ENTER now
<committable6> Zoffix__, ¦HEAD(1ea3ff2): «0.7404541␤»
```

Reply:
> if an atomicintis needed for a 0-or-1 cas switch

It isn't. `cas` can work with arbitrary objects: https://github.com/perl6/doc/issues/2393

Reply:
I do like the ```when``` aesthetically, but I changed it to ```if```/```else``` in the most recent commit due to the performance difference you pointed out.